### PR TITLE
[Datasource, Resource] Network Interface/s

### PR DESCRIPTION
**Pull request by michaelkad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
=== RUN   TestAccIBMPINetworkInterfacesDataSourceBasic
--- PASS: TestAccIBMPINetworkInterfacesDataSourceBasic (19.35s)
PASS

=== RUN   TestAccIBMPINetworkInterfaceDataSourceBasic
--- PASS: TestAccIBMPINetworkInterfaceDataSourceBasic (16.10s)
PASS

=== RUN   TestAccIBMPINetworkInterfaceBasic
--- PASS: TestAccIBMPINetworkInterfaceBasic (21.81s)
PASS

=== RUN   TestAccIBMPINetworkInterfaceAllArgs
--- PASS: TestAccIBMPINetworkInterfaceAllArgs (30.83s)
PASS
...
```

### CODE REVIEW

File: ibm/service/power/data_source_ibm_pi_network_interface.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	d.Set(Attr_NetworkSecurityGroupID, networkInterface.NetworkSecurityGroupID)
+
+	if networkInterface.Instance != nil {
+		pvmInstance := []map[string]interface{}{}
+		instanceMap := pvmInstanceToMap(networkInterface.Instance)
+		pvmInstance = append(pvmInstance, instanceMap)
+		d.Set(Attr_PVMInstance, pvmInstance)
+	}
+	d.Set(Attr_Status, networkInterface.Status)
+	if len(networkInterface.UserTags) > 0 {
```
Comment (Alexander-Kita) [inline ibm/service/power/data_source_ibm_pi_network_interface.go:127]: This is setting Attr_UserTags but I do not see it in the schema

File: website/docs/r/pi_network_interface.html.markdown [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+## Argument Reference
+
+Review the argument references that you can specify for your resource.
+
+- `pi_cloud_instance_id` - (Required, Forces new resource, String) The GUID of the service instance associated with an account.
+- `pi_instance_id` - (Optional, String) If supplied populated it attaches to the PVMInstanceID, if empty detaches from PVMInstanceID.
+- `pi_ip_address` - (Optional,Forces new resource, String) The requested IP address of this network interface.
+- `pi_name` - (Optional, String) Name of the network interface.
+- `pi_network_id` - (Required, Forces new resource, String) network id.
+- `pi_user_tags` - (Optional, Forces new resource, List) The user tags attached with this resource.
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_interface.html.markdown:48]:
```suggestion
- `pi_user_tags` - (Optional, Forces new resource, List) The user tags attached to this resource.
```

Slight wording fix, may need to be changed throughout.

File: ibm/service/power/data_source_ibm_pi_network_interface.go (hunk 1 above)
Comment (michaelkad) [reply ibm/service/power/data_source_ibm_pi_network_interface.go]: Removed.

File: website/docs/r/pi_network_interface.html.markdown (hunk 2 above)
Comment (michaelkad) [reply website/docs/r/pi_network_interface.html.markdown]: Updated

File: ibm/service/power/data_source_ibm_pi_instances.go [hunk 3]
```diff
@@ -273,7 +273,7 @@ func flattenPvmInstanceNetworks(list []*models.PVMInstanceNetwork) (networks []m
 			p := make(map[string]interface{})
 			p[Attr_ExternalIP] = pvmip.ExternalIP
 			p[Attr_IP] = pvmip.IPAddress
-			p[Attr_MacAddress] = pvmip.MacAddress
+			p[Attr_Macaddress] = pvmip.MacAddress
```
Comment (ismirlia) [inline ibm/service/power/data_source_ibm_pi_instances.go:276]: isn't Attr_Macaddress deprecated? should it be Attr_MacAddress?

Comment (michaelkad) [reply ibm/service/power/data_source_ibm_pi_instances.go:276]: Good catch! Both need to set here, will be making another PR to remove them later.

File: ibm/service/power/ibm_pi_constants.go [hunk 4]
```diff
@@ -4,6 +4,9 @@ import "time"
 
 const (
 	// Arguments
+	Arg_NetworkSecurityGroupID              = "pi_network_security_group_id"
+	Arg_UserTags                            = "pi_user_tags"
+	Arg_IPAddress                           = "pi_ip_address"
```
Comment (Alexander-Kita) [inline ibm/service/power/ibm_pi_constants.go:9]: These need to be put into alphabetical order

Comment (michaelkad) [reply ibm/service/power/ibm_pi_constants.go]: Updated

File: ibm/service/power/resource_ibm_pi_network_interface.go [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+			},
+			Arg_NetworkID: {
+				Description:  "Network ID.",
+				ForceNew:     true,
+				Required:     true,
+				Type:         schema.TypeString,
+				ValidateFunc: validation.NoZeroValues,
+			},
+			Arg_UserTags: {
+				Description: "The user tags attached with this resource.",
```
Comment (Alexander-Kita) [inline ibm/service/power/resource_ibm_pi_network_interface.go:70]:
```suggestion
				Description: "The user tags attached to this resource.",
```

Comment (michaelkad) [reply ibm/service/power/resource_ibm_pi_network_interface.go]: updated

**Review by Alexander-Kita — APPROVED** [review]

**Comment by michaelkad** [comment]
Move to main repo.

**Pull request closed without merging**